server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -3378,7 +3378,10 @@ public void addFlushListener(Translog.Location location, ActionListener<Long> li
         this.flushListener.addOrNotify(location, new ActionListener<>() {
             @Override
             public void onResponse(Long generation) {
-                waitForCommitDurability(generation, listener.map(v -> generation));
+                // No need to wait for commit durability since the ONLY caller is PostWriteRefresh which
+                // requires the commit generation to be visible to search shards. Visibility does not require
+                // special logic for durability since nodes can talk to each other directly (RCO).
+                notWaitForCommitDurability(generation, listener.map(v -> generation));
```

Comment:
This change means `waitForCommitDurability` is only called inside `flushHoldingLock`. I think it's an improvement since it is much easier to reason about the threading and we use thread local to control whether waiting for upload is relevant.

Reply:
It makes sense to me. Can we add javadoc on `Engine.addFlushListener` (or perhaps on the method locally here) to state that it only waits for the flush, not commit durability and if this is desired it must be handled by the caller?

Reply:
Good point on Javadoc. Added in 16c6350f8dcd6dae84e9a4b38d81c613117dfe4d